// notebooklm-client/dist/index.js
/**
 * notebooklm-client — Standalone NotebookLM API client.
 *
 * @example
 * ```ts
 * import { NotebookClient } from 'notebooklm-client';
 *
 * // Browser mode (default — launches Chrome)
 * const client = new NotebookClient();
 * await client.connect();
 *
 * // HTTP mode (no browser — uses exported session + Chrome-like TLS)
 * const client2 = new NotebookClient();
 * await client2.connect({ transport: 'http' });
 * ```
 */
export { NotebookClient } from './client.js';
export { BrowserTransport } from './transport-browser.js';
export { HttpTransport } from './transport-http.js';
export { CurlTransport } from './transport-curl.js';
export { TlsClientTransport } from './transport-tlsclient.js';
export { detectBestTier, createTransport, TIER_LABELS } from './transport-resolver.js';
// Paths (multi-account support)
export { getHomeDir, setHomeDir, getSessionPath, getProfileDir, getRpcIdsPath } from './paths.js';
// Session persistence
export { saveSession, loadSession, hasValidSession, refreshTokens } from './session-store.js';
// TLS config
export { CHROME_CIPHERS, CHROME_SIGALGS, chromeTlsOptions } from './tls-config.js';
export { SessionError, BrowserError } from './errors.js';
export { parseEnvelopes, stripSafetyPrefix } from './boq-parser.js';
